File: starterkit/java/StateManager.java
import java.util.function.Consumer;

public class StateManager {
    Consumer<Boolean> m_curfunc;
    Consumer<Boolean> m_nextfunc;
    Consumer<Boolean> m_tempfunc;

    Boolean           m_noWait;

    public void Update() {
        while(true) {
            Boolean bFirst = false;
            if (m_nextfunc != null) {
                m_curfunc = m_nextfunc;
                m_nextfunc = null;
                bFirst = true;
            }
            m_noWait = false;
            if (m_curfunc!=null) {
                m_curfunc.accept(bFirst);
            }
            if (!m_noWait) break;
        }
    }

    public void Goto(Consumer<Boolean> func) {
        m_nextfunc = func;
    }

    public Boolean CheckState(Consumer<Boolean> func) {
        return (m_curfunc == func);
    }

    public void SetNextState(Consumer<Boolean> func) {
        m_tempfunc = func;
    }
    public void GoNextState() {
        m_nextfunc = m_tempfunc;
        m_tempfunc = null;
    }
    public Boolean HasNextState() {
        return m_tempfunc != null;
    }
    public void NoWait()
    {
        m_noWait = true;
    }
}

File: starterkit/php/StateManager.php
<?php
abstract class StateManager {
    abstract protected function call_state($state, $b);

    private $m_curfunc  = NULL;
    private $m_nextfunc = NULL;

    private $m_noWait   = FALSE;

    public function Update() {
        while(TRUE) {
            $bFirst = FALSE;
            if ($this->m_nextfunc!=NULL) {
                $this->m_curfunc = $this->m_nextfunc;
                $this->m_nextfunc = NULL;
                $bFirst = TRUE;
            }
            $this->m_noWait = FALSE;
            if ($this->m_curfunc!=NULL) {
                $this->call_state($this->m_curfunc, $bFirst);
            }
            if ($this->m_noWait == FALSE) {
                break;
            }
        }
    }

    public function GotoState($st) {
        $this->m_nextfunc = $st;
    }

    public function CheckState($st) {
        return $st == $this->m_curfunc;
    }
 
    public function HasNextState() {
        return $this->m_nextfunc != NULL;
    }

    public function NoWait() {
        $this->m_noWait = TRUE;
    }
}
?>